File: pkg/dart2cpp/sample/cpp/closure_capture_test.cpp
#include "dart2cpp.h"

// 工具宏定义

// ============================================================================
// 类: Calculator
// ============================================================================

class Calculator {
public:
  Int base;
  Calculator(Int base) : base(base) {
  }
  
  ObjectPtr<Function> makeAdder() {
    return makeFunction([&](Int x) { return this->base->operator_add(x); });
  }
  
};

Nullable testBasicClosure();
Nullable testMultipleCapture();
Nullable testMemberFunctionCapture();
Nullable testClosureInLoop();
Nullable testNestedClosure();
Nullable testBasicClosure() {
  dart_print(dart_string("Test 1: Basic Closure"));
auto x = dart_int(10);
auto addX = makeFunction([&](Int y) { return x->operator_add(y); }, std::vector<Any>{Any(x)});
dart_print(addX->call(dart_int(5)));
return Void;
}

Nullable testMultipleCapture() {
  dart_print(dart_string("Test 2: Multiple Variable Capture"));
auto a = dart_int(5);
auto b = dart_int(10);
auto prefix = dart_string("Result: ");
auto compute = makeFunction([&]() { auto sum = a->operator_add(b);
return prefix->operator_add(sum->toString()); }, std::vector<Any>{Any(a), Any(b), Any(prefix)});
dart_print(compute->call());
return Void;
}

Nullable testMemberFunctionCapture() {
  dart_print(dart_string("Test 3: Member Function Capture"));
auto calc = ObjectPtr<Calculator>(new Calculator(dart_int(100)));
auto adder = calc->makeAdder();
dart_print(adder->call(dart_int(23)));
return Void;
}

Nullable testClosureInLoop() {
  dart_print(dart_string("Test 4: Closure in Loop"));
auto functions = _GrowableList::create(dart_int(0));
for (auto i = dart_int(0); i->operator_less(dart_int(3)); i = i->operator_add(dart_int(1))) {
functions->add(makeFunction([&]() { return i; }, std::vector<Any>{Any(i)}));
}
auto sync_for_iterator = functions->iterator();
for (; sync_for_iterator->hasNext(); ) {
auto f = sync_for_iterator->next();
dart_print(f->call());
};
return Void;
}

Nullable testNestedClosure() {
  dart_print(dart_string("Test 5: Nested Closure"));
auto outer = dart_int(1);
auto outerFunc = makeFunction([&]() { auto middle = dart_int(10);
auto innerFunc = makeFunction([&]() { return outer->operator_add(middle); }, std::vector<Any>{Any(outer), Any(middle)});
return innerFunc->call(); });
dart_print(outerFunc->call());
return Void;
}

// ============================================================================
// 主函数
// ============================================================================

int main() {
  try {
    testBasicClosure();
testMultipleCapture();
testMemberFunctionCapture();
testClosureInLoop();
testNestedClosure();
    return 0;
  } catch (const std::exception& e) {
    std::cerr << "Error: " << e.what() << std::endl;
    return 1;
  }
}

File: pkg/dart2cpp/sample/dart/closure_capture_test.dart
/// 闭包变量捕获功能测试

void main() {
  // 测试1: 基础闭包 - 捕获局部变量
  testBasicClosure();

  // 测试2: 闭包捕获多个变量
  testMultipleCapture();

  // 测试3: 成员函数引用 - 捕获this
  testMemberFunctionCapture();

  // 测试4: 闭包在循环中
  testClosureInLoop();

  // 测试5: 嵌套闭包
  testNestedClosure();
}

// 测试1: 基础闭包捕获
void testBasicClosure() {
  print("Test 1: Basic Closure");

  int x = 10;

  var addX = (int y) {
    return x + y; // 捕获外部变量x
  };

  print(addX(5)); // 应该输出15
}

// 测试2: 捕获多个变量
void testMultipleCapture() {
  print("Test 2: Multiple Variable Capture");

  int a = 5;
  int b = 10;
  String prefix = "Result: ";

  var compute = () {
    int sum = a + b; // 捕获a和b
    return prefix + sum.toString(); // 捕获prefix
  };

  print(compute()); // 应该输出"Result: 15"
}

// 测试3: 成员函数捕获
class Calculator {
  int base;

  Calculator(this.base);

  Function makeAdder() {
    // 这个闭包应该捕获this
    return (int x) => base + x;
  }
}

void testMemberFunctionCapture() {
  print("Test 3: Member Function Capture");

  var calc = Calculator(100);
  var adder = calc.makeAdder();
  print(adder(23)); // 应该输出123
}

// 测试4: 循环中的闭包
void testClosureInLoop() {
  print("Test 4: Closure in Loop");

  List<Function> functions = [];

  for (int i = 0; i < 3; i++) {
    // 每个闭包应该捕获当前的i值
    functions.add(() => i);
  }

  // 注意：Dart的for循环会为每次迭代创建新的i变量
  for (var f in functions) {
    print(f());
  }
}

// 测试5: 嵌套闭包
void testNestedClosure() {
  print("Test 5: Nested Closure");

  int outer = 1;

  var outerFunc = () {
    int middle = 10;

    var innerFunc = () {
      return outer + middle; // 捕获outer和middle
    };

    return innerFunc();
  };

  print(outerFunc()); // 应该输出11
}
